### ISSUE TITLE
Error related to sourcemaps in production build

### ISSUE THREAD

**Issue by IgnisDa** [remix-run/remix#8309]
### Reproduction

Using a unstable vite template and building and deploying it gives an error about a sourcemap file missing.

All remix packages are at `v2.4.0`.

[Here](https://github.com/IgnisDa/ryot/blob/2801f78d73cdf0bf81d77740aef28d05cc652e3f/Dockerfile#L56) is the part of the dockerfile that copies relevant files.

### System Info

```shell
System:
    OS: Linux 6.6 Arch Linux
    CPU: (16) x64 AMD Ryzen 7 5800H with Radeon Graphics
    Memory: 19.07 GB / 27.26 GB
    Container: Yes
    Shell: 5.2.15 - /bin/bash
  Binaries:
    Node: 20.5.1 - ~/.proto/bin/node
    Yarn: 4.0.0 - ~/.proto/bin/yarn
    npm: 10.2.1 - /usr/sbin/npm
    bun: 1.0.7 - ~/.proto/bin/bun
```

### Used Package Manager

yarn

### Expected Behavior

The production server to work fine.

### Actual Behavior

Here is the stderr for the server:
```
Error: ENOENT: no such file or directory, open '/home/ryot/build/server/index.js.map'
    at readFileSyncUtf8 (node:internal/fs/read/utf8:20:3)
    at Object.readFileSync (node:fs:467:12)
    at Array.retrieveSourceMap (/home/ryot/node_modules/@remix-run/serve/dist/cli.js:49:37)
    at /home/ryot/node_modules/source-map-support/source-map-support.js:85:24
    at mapSourcePosition (/home/ryot/node_modules/source-map-support/source-map-support.js:216:21)
    at wrapCallSite (/home/ryot/node_modules/source-map-support/source-map-support.js:397:20)
    at Function.prepareStackTrace (/home/ryot/node_modules/source-map-support/source-map-support.js:446:39)
    at maybeOverridePrepareStackTrace (node:internal/errors:140:29)
    at prepareStackTrace (node:internal/errors:114:5)
    at getStackString (node:internal/util/inspect:1239:16)
```

**Comment by VHall1** [comment]
Can confirm, I'm having the same issue on Windows 11.

As far as I can tell, it seems like this is related to the [2.4.0 release](https://github.com/remix-run/remix/releases/tag/remix%402.4.0). Pegged all remix packages locally to version 2.3.1 and that seems to have "solved" the issue.

Think the issue might be related to this new feature: https://github.com/remix-run/remix/pull/8174

Stack trace for reference:
```
Error: ENOENT: no such file or directory, open 'C:\C:\Users\Frk\onx\onx-mod\apps\website\build\index.js.map'
    at Object.readFileUtf8 (node:internal/fs/sync:25:18)
    at Object.readFileSync (node:fs:441:19)
    at Array.retrieveSourceMap (C:\Users\Frk\onx\onx-mod\node_modules\@remix-run\serve\dist\cli.js:49:37)
    at C:\Users\Frk\onx\onx-mod\node_modules\source-map-support\source-map-support.js:85:24
    at mapSourcePosition (C:\Users\Frk\onx\onx-mod\node_modules\source-map-support\source-map-support.js:216:21)
    at wrapCallSite (C:\Users\Frk\onx\onx-mod\node_modules\source-map-support\source-map-support.js:397:20)
    at Function.prepareStackTrace (C:\Users\Frk\onx\onx-mod\node_modules\source-map-support\source-map-support.js:446:39)
    at maybeOverridePrepareStackTrace (node:internal/errors:141:29)
    at prepareStackTrace (node:internal/errors:115:5)
    at process.emit (C:\Users\Frk\onx\onx-mod\node_modules\source-map-support\source-map-support.js:508:52) {
  errno: -4058,
  code: 'ENOENT',
  syscall: 'open',
  path: 'C:\\C:\\Users\\Frk\\onx\\onx-mod\\apps\\website\\build\\index.js.map'
}
```

**Comment by IgnisDa** [comment]
@VHall1 Does this problem happen for you in docker or normal builds?

**Comment by VHall1** [comment]
Normal builds. Got a repo to reproduce the issue going in here: https://github.com/VHall1/dummy-remix-app. Thrown an error inside of the index route and got the same issue

![image](https://github.com/remix-run/remix/assets/46335285/461a03c5-5e5d-445f-8b89-10f63f4f7d51)

**Comment by IgnisDa** [comment]
THanks for the repro!

**Comment by kiliman** [comment]
I'll take a look. Thanks for the report.

**Comment by VHall1** [comment]
Just tested on mac and that seems to work, so sounds like this is a platform issue

**Comment by VHall1** [comment]
Took a stab at submitting a PR to fix this issue. Would appreciate another pair of eyes to make sure everything makes sense as this is my first contribution to the project.

**Comment by brophdawg11** [comment]
This is resolved by #8321 and will be included in the next release 👍

**Comment by adaboese** [comment]
Confirming that pegging to v2.3.1 fixes the issue. Thank you

**Comment by IgnisDa** [comment]
@Hey @brophdawg11, I tried this wil `2.4.1-pre.2`, but I get the same error.

**Comment by kiliman** [comment]
Are you using Remix App Server or Express?

Can you log the `source` and `filePath` in the `retrieveSourceMap` handler?

https://github.com/remix-run/remix/pull/8321/files?diff=unified&w=1

It looks like this fix works for macOS and Windows, but perhaps Linux is different. Does the map file actually exist at the paths logged?

**Comment by IgnisDa** [comment]
> Are you using Remix App Server or Express?

The default one provided in the vite template.

> Does the map file actually exist at the paths logged?

It does not.

**Comment by kiliman** [comment]
By default, Vite doesn't generate sourcemaps for production builds. You will need to pass the `--sourcemap` argument to the build command.

Just note that sourcemaps contain the _entire_ source, so client sourcemaps will contain server code as well.

**Comment by fobos531** [comment]
> By default, Vite doesn't generate sourcemaps for production builds. You will need to pass the `--sourcemap` argument to the build command.
> 
> Just note that sourcemaps contain the _entire_ source, so client sourcemaps will contain server code as well.

Hello @kiliman , is it known why the `remix-serve` expects to find this sourcemap file in the first place? I imagine we would not want to build and deploy the app with sourcemaps in the build output as well.

**Comment by IgnisDa** [comment]
Yep, as mentioned above, I do not want to generate the final build with sourcemaps (or anything useless for that matter).

**Comment by kiliman** [comment]
@IgnisDa Hmm... then I'm not sure why it's trying to access sourcemaps if you're not generating them. Can you check `build/index.js` to see if it has the comment at the end?
```ts
//# sourceMappingURL=index.js.map
```
If not, I'm not sure why `source-map-support` is trying to load the sourcemap then. Do we need to add a `fs.existsSync()` check after `url.fileURLToPath()`?

@brophdawg11 we may need to re-open this.

**Comment by fobos531** [comment]
> @IgnisDa Hmm... then I'm not sure why it's trying to access sourcemaps if you're not generating them. Can you check `build/index.js` to see if it has the comment at the end?
> 
> ```ts
> //# sourceMappingURL=index.js.map
> ```
> 
> If not, I'm not sure why `source-map-support` is trying to load the sourcemap then. Do we need to add a `fs.existsSync()` check after `url.fileURLToPath()`?
> 
> @brophdawg11 we may need to re-open this.

Hello @kiliman 

Just checked on my end - to give more context, this is a Vite build. I do NOT have this in the generated `build/server/index.js`. So it definitely looks like some part of code is trying to load the sourcemap while it doesn't at all exist.

**Comment by adaboese** [comment]
FYI Upgrading to v2.4.1 did not fix the issue.

**Comment by kiliman** [comment]
Is this only happening in Docker? Linux?

I cannot reproduce this using Vite Dev Server or Custom Express server on macOS.

**Comment by kiliman** [comment]
Also, does the error occur on startup, and your app won't run? Or do you get the sourcemap error when an exception is thrown?

**Comment by adaboese** [comment]
This is happening only in Docker when the container is started.

**Comment by kiliman** [comment]
I'm using Fly.io, which also uses Docker containers. 

https://github.com/kiliman/remix-vite-template

<img width="1257" alt="image" src="https://github.com/remix-run/remix/assets/47168/4a899eac-565a-4b2b-b6f7-224005170ed5">

https://remix-vite-template.fly.dev/

Sorry I can't be of more help.

**Comment by IgnisDa** [comment]
@kiliman can you commit your dockerfile too in the above repo?

**Comment by kiliman** [comment]
The Dockerfile is in the remix-vite-template repo.

But here's an even simpler repo. It's literally just the remix-vite unstable template with a Dockerfile. Just run `docker compose up` and you should be able to open http://localhost:3000

https://github.com/kiliman/remix-docker-test

<img width="1168" alt="image" src="https://github.com/remix-run/remix/assets/47168/dddacc73-0d57-46ab-a3ed-aec373557f2b">

**Comment by IgnisDa** [comment]
I tried with the new scripts in your package.json. It still does not work in a multi stage docker build.

**Comment by kiliman** [comment]
Can you create a simpler setup? The project you linked to has a lot of tools I'm not familiar with: moon and rust.

I definitely want to figure out what's causing the issue for you.

Thanks!

**Comment by IgnisDa** [comment]
Hey @kiliman I was able to reproduce the error in the repo you linked above https://github.com/IgnisDa/remix-docker-test.

![image](https://github.com/remix-run/remix/assets/60938164/efc26773-3876-46fc-8180-0c3467671848)

**Comment by kiliman** [comment]
Ok, I was able to reproduce it. Re-opening this issue.

**Comment by kiliman** [comment]
Here's a patch while I work on the PR

`patches/@remix-run+serve+2.4.1.patch`
```diff
diff --git a/node_modules/@remix-run/serve/dist/cli.js b/node_modules/@remix-run/serve/dist/cli.js
index fa63111..6647eea 100755
--- a/node_modules/@remix-run/serve/dist/cli.js
+++ b/node_modules/@remix-run/serve/dist/cli.js
@@ -41,13 +41,15 @@ var getPort__default = /*#__PURE__*/_interopDefaultLegacy(getPort);
 process.env.NODE_ENV = process.env.NODE_ENV ?? "production";
 sourceMapSupport__default["default"].install({
   retrieveSourceMap: function (source) {
-    // get source file with the `file://` prefix
-    let match = source.match(/^file:\/\/(.*)$/);
-    if (match) {
-      let filePath = url__default["default"].fileURLToPath(source);
+    let filePath = source;
+    if (filePath.startsWith("file://")) {
+      filePath = url__default["default"].fileURLToPath(source);
+    }
+    let mapPath = `${filePath}.map`;
+    if (fs__default["default"].existsSync(mapPath)) {
       return {
         url: source,
-        map: fs__default["default"].readFileSync(`${filePath}.map`, "utf8")
+        map: fs__default["default"].readFileSync(mapPath, "utf8")
       };
     }
     return null;
```

**Comment by adaboese** [comment]
Not sure what's the difference, but that patch didn't quite work for me. This one did:

```diff
diff --git a/dist/cli.js b/dist/cli.js
index fa631116d949ac08887e5a6e7a0a62c5107e77a3..6647eea016df7334db87e973ea1e6bfb933b744c 100644
--- a/dist/cli.js
+++ b/dist/cli.js
@@ -41,13 +41,15 @@ var getPort__default = /*#__PURE__*/_interopDefaultLegacy(getPort);
 process.env.NODE_ENV = process.env.NODE_ENV ?? "production";
 sourceMapSupport__default["default"].install({
   retrieveSourceMap: function (source) {
-    // get source file with the `file://` prefix
-    let match = source.match(/^file:\/\/(.*)$/);
-    if (match) {
-      let filePath = url__default["default"].fileURLToPath(source);
+    let filePath = source;
+    if (filePath.startsWith("file://")) {
+      filePath = url__default["default"].fileURLToPath(source);
+    }
+    let mapPath = `${filePath}.map`;
+    if (fs__default["default"].existsSync(mapPath)) {
       return {
         url: source,
-        map: fs__default["default"].readFileSync(`${filePath}.map`, "utf8")
+        map: fs__default["default"].readFileSync(mapPath, "utf8")
       };
     }
     return null;
```

and in your `package.json`:

```
"@remix-run/serve@2.4.1": "patches/@remix-run__serve@2.4.1.patch"
```

**Comment by IgnisDa** [comment]
Hey @kiliman, any updates on this?

**Comment by TaQuanMinhLong** [comment]
I got the same problem when trying to throw zod error into the catch block, but it wasn't thrown, instead the remix-serve yelling that the can't find zod's source map, I did try build in production mode but still get the same error thrown, except no stack strace, why do we even include the source map support in production tho? And at the end, my current temporary solution is to comment out the source map support install part from the remix-serve package in node modules, and it works as normal

**Comment by justinwaite** [comment]
Maybe this will help someone else, but I was getting this error when `@libsql/client` was being bundled in the server build. Externalizing it (`ssr.external = ['libsql', '@libsql/client']`) in the vite config fixed my issue.

**Comment by swalker326** [comment]
Just to add some more discussion/a pnpm solution

I've fixed it by applying a patch to remix. 

`pnpm patch @remix-run/serve`

it will prompt you with a file to edit, here is the new content (if on mac you can cmd +click it to open it in vscode)
```ts
sourceMapSupport__default["default"].install({
  retrieveSourceMap: function (source) {
    // get source file with the `file://` prefix
    let match = source.match(/^file:\/\/(.*)$/);
    if (match) {
      let filePath = url__default["default"].fileURLToPath(source);
      //   return {
      //     url: source,
      //     map: fs__default["default"].readFileSync(`${filePath}.map`, "utf8")
      //   };
    }
    return null;
  }
});
```

I just commented out the return so it always is returning null on sourceMap

In the same terminal where you ran patch it should give you a command to run to apply the patch. Something like this:
`pnpm patch-commit /private/var/folders/q7/qmr04v0566bbsq7g4fzhr6tr0000gn/T/38893e41339ff1ba3643d6010ef1f544`


You also have to update your docker file to copy over your patches directory at build time so pnpm install will apply them.
```
# Install node modules
COPY --link patches patches/
```

Such is life in unstable land.

**Comment by Mxs2019** [comment]
Any updates here. I'm getting this error. Is this the same bug? 

```
Error: ENOENT: no such file or directory, open '/app/node_modules/.pnpm/axios@1.6.0/node_modules/axios/lib/adapters/http.js.map'
2024-01-03 11:10:33.565 [fly_io] [INFO] ewr windmill-staging at Object.openSync (node:fs:596:3)
2024-01-03 11:10:33.565 [fly_io] [INFO] ewr windmill-staging at Object.readFileSync (node:fs:464:35)
2024-01-03 11:10:33.565 [fly_io] [INFO] ewr windmill-staging at Array.retrieveSourceMap (/app/node_modules/.pnpm/@remix-run+serve@2.4.1_typescript@5.2.2/node_modules/@remix-run/serve/dist/cli.js:50:37)
2024-01-03 11:10:33.565 [fly_io] [INFO] ewr windmill-staging at /app/node_modules/.pnpm/source-map-support@0.5.21/node_modules/source-map-support/source-map-support.js:85:24
2024-01-03 11:10:33.565 [fly_io] [INFO] ewr windmill-staging at mapSourcePosition (/app/node_modules/.pnpm/source-map-support@0.5.21/node_modules/source-map-support/source-map-support.js:216:21)
2024-01-03 11:10:33.565 [fly_io] [INFO] ewr windmill-staging at wrapCallSite (/app/node_modules/.pnpm/source-map-support@0.5.21/node_modules/source-map-support/source-map-support.js:397:20)
2024-01-03 11:10:33.565 [fly_io] [INFO] ewr windmill-staging at Function.prepareStackTrace (/app/node_modules/.pnpm/source-map-support@0.5.21/node_modules/source-map-support/source-map-support.js:446:39)
2024-01-03 11:10:33.565 [fly_io] [INFO] ewr windmill-staging at maybeOverridePrepareStackTrace (node:internal/errors:140:29)
2024-01-03 11:10:33.565 [fly_io] [INFO] ewr windmill-staging at prepareStackTrace (node:internal/errors:114:5)
2024-01-03 11:10:33.565 [fly_io] [INFO] ewr windmill-staging at parseStackFrames (/app/node_modules/.pnpm/@sentry+utils@7.70.0/node_modules/@sentry/src/eventbuilder.ts:22:1)
2024-01-03 11:10:33.565 [fly_io] [INFO] ewr windmill-staging at exceptionFromError (/app/node_modules/.pnpm/@sentry+utils@7.70.0/node_modules/@sentry/src/eventbuilder.ts:34:1)
2024-01-03 11:10:33.565 [fly_io] [INFO] ewr windmill-staging at Object.eventFromUnknownInput (/app/node_modules/.pnpm/@sentry+utils@7.70.0/node_modules/@sentry/src/eventbuilder.ts:103:1)
2024-01-03 11:10:33.565 [fly_io] [INFO] ewr windmill-staging at NodeClient.eventFromException (/app/node_modules/.pnpm/@sentry+core@7.70.0/node_modules/@sentry/src/server-runtime-client.ts:52:1)
2024-01-03 11:10:33.565 [fly_io] [INFO] ewr windmill-staging at NodeClient.captureException (/app/node_modules/.pnpm/@sentry+core@7.70.0/node_modules/@sentry/src/baseclient.ts:157:1)
2024-01-03 11:10:33.565 [fly_io] [INFO] ewr windmill-staging at NodeClient.captureException (/app/node_modules/.pnpm/@sentry+core@7.70.0/node_modules/@sentry/src/server-runtime-client.ts:87:1)
2024-01-03 11:10:33.565 [fly_io] [INFO] ewr windmill-staging at /app/node_modules/.pnpm/@sentry+core@7.70.0/node_modules/@sentry/src/hub.ts:200:1 {
2024-01-03 11:10:33.565 [fly_io] [INFO] ewr windmill-staging errno: -2,
2024-01-03 11:10:33.565 [fly_io] [INFO] ewr windmill-staging syscall: 'open',
2024-01-03 11:10:33.565 [fly_io] [INFO] ewr windmill-staging code: 'ENOENT',
2024-01-03 11:10:33.565 [fly_io] [INFO] ewr windmill-staging path: '/app/node_modules/.pnpm/axios@1.6.0/node_modules/axios/lib/adapters/http.js.map'
2024-01-03 11:10:33.565 [fly_io] [INFO] ewr windmill-staging }
```

**Comment by IgnisDa** [comment]
The most straight-forward solution right now is to stay on 2.3.1 or apply one of the patches recommended above.

**Comment by VHall1** [comment]
Looking at this again, and wondering if one of the issues only happens when errors are thrown by external packages, rather than the components themselves. Seems like source-map-support might be passing the dependency URL in the `source` as part of the stack trace, and because we don't have sourcemaps available for those dependencies it fails. Maybe a strawman would be to fallback to the unfriendly path if no sourcemaps are available?

**Comment by dwcarr** [comment]
I am running 2.4.1, base remix server, no vite, and I am getting a very similar error, only related to zod, though.

<img width="1107" alt="Screenshot 2024-01-06 at 9 32 25 PM" src="https://github.com/remix-run/remix/assets/8039670/30966e35-8939-4c8d-b0d8-ed6ba6331390">

**Comment by ledgerfetcher** [comment]
Experienced the same with [postgres.js](https://github.com/porsager/postgres) ✌️ Solved the issue by downgrading to `2.3.1`.

**Comment by dwcarr** [comment]
Yes, I can confirm that downgrading to 2.3.1 fixes the issue.

**Comment by VHall1** [comment]
Gave this another shot. I believe this should resolve both of the issues raised above.

**Comment by brophdawg11** [comment]
Thanks @VHall1!  This should be resolved by #8446 and available in the next release.

**Comment by IgnisDa** [comment]
It works with the above pre-release. Thanks @VHall1!

**Comment by VHall1** [comment]
that is great! Thanks for confirming 😄

**Comment by IgnisDa** [comment]
Hey @VHall1, when will `v2.5.0` be released?

**Comment by erlangparasu** [comment]
any updates?

**Comment by IgnisDa** [comment]
@erlangparasu This has long since been fixed with 2.5.0. I recommend you upgrade to that version and if the issue still persists, create a new issue.

**Comment by erlangparasu** [comment]
> @erlangparasu This has long since been fixed with 2.5.0. I recommend you upgrade to that version and if the issue still persists, create a new issue.

thanks 2.5.0 works
